pkg/controller/compute/droplet.go

```diff
@@ -87,9 +88,16 @@ func (c *dropletExternal) Observe(ctx context.Context, mg resource.Managed) (man
 	if !ok {
 		return managed.ExternalObservation{}, errors.New(errNotDroplet)
 	}
-	observed, _, err := c.Droplets.Get(ctx, cr.Status.AtProvider.ID)
+	externalID, err := strconv.Atoi(cr.ObjectMeta.Annotations[meta.AnnotationKeyExternalName])
 	if err != nil {
-		return managed.ExternalObservation{}, errors.Wrap(err, errGetDroplet)
+		// on the first try the value of 'crossplane.io/external-name' annotaion
```

Comment:
An initializer is actually setting this to the name of the CR. You can disable it by manually providing the initializers list so that you don't need to handle this case, see example here https://github.com/crossplane/provider-aws/blob/087f587/pkg/controller/ec2/vpc/controller.go#L66

Then you can check whether `meta.GetExternalName(cr)` returns empty string, if so, it'd mean resource does not exist since `Create` hasn't set it yet.

Reply:
Good to know, I'll look into it shortly, but a follow up question specifically on this topic. I feel like what I'm doing here (i.e. using `external-name` annotation to store numeric ID of the managed resource) is anti-pattern, at the very least feels a bit weird to store _ID_ in a field called _name_. What do you think, is this good approach after all or too hacky workaround?

Worth mentioning that the only reason behind this is to overcome the (somewhat) limitation of `godo` library which doesn't expose searching by name functionality.

Reply:
The main point is that external-name + required fields should be [enough](https://github.com/crossplane/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#external-resource-name) to do all of the operations. Using a field in the status is not an option since status [is considered to be ephemeral](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status).

However, I know that there are resources that expose name but query with a string that's generated to satisfy the specific format of the query. So, it does not have to be exact. In this case though, ID is completely separate and not something that can be generated by looking at other information in spec or metadata. From that perspective, it's pretty similar VPC in AWS; there is no way to name a VPC, you only get an ID when you [create](https://github.com/crossplane/provider-aws/blob/087f587/pkg/controller/ec2/vpc/controller.go#L177). Also `SecurityGroup` has both name and ID but we use ID since that's what's required in most of the calls. Also we use the ID in external-name so that importing an existing resource works as expected. `external-name` in that sense could have been named as `external-id` as well but then we'd be in the same position again, there are lots of resources who use `name` instead of ID. So, we chose the more humane one, i.e. `name`.

To sum up, there is some room for liberty as long as it meets those expectations from external name annotation. I'd also lean towards using the name instead of ID but it needs to work when `status` is gone and allow importing existing resource.

Reply:
Thank you @muvaf for the very detailed explanation! Was there any reason (or discussion for that matter) to define mutually exclusive `external-name` and `external-id` in which `external-name` is the default behavior Crossplane core looks up to and `external-id` can be used in edge cases such as this and AWS VPC and etc as a fallback mechanism?

Anyways, I've applied your suggestions, if the current state of PR looks good please let me know to squash and cleanup my commits and we can proceed with merging this.

Reply:
> Was there any reason (or discussion for that matter) to define mutually exclusive external-name and external-id in which external-name is the default behavior Crossplane core looks up to and external-id can be used in edge cases such as this and AWS VPC and etc as a fallback mechanism?

Honestly, I don't remember us discussing the edge case where you have both name and the ID. So, we decided whatever is the identifier it should go to `external-name`.

I think the PR is good to go 🙂